[KongHeyi] iP

heeeyi opened the pull request:
## Duke Plus

> Software and cathedrals are much the same; first we build them, then we pray. – by Anonymous

Duke Plus is a task-management app, with the attentive design of saving your memory losses out of long-term intensive coding. If you are also worried about yourself forgetting things, click [here](https://xkcd.com/1093/) to find some consolations.


Aside from the fact that you yourself need to go through stressful coding to produce it, as a user, you will feel it 

* *Easy* to use 😄 
* Friendly to **typewriting-lovers**
* ~~Not friendly to people too lazy to type the task~~

What you need to do to use it is,
1. Download it (coming soon)
2. Double-click to run
3. Type your tasks to add
4. Ask it to remind you when tasks are coming up

Features:

- [x] Adding, deleting, marking tasks
- [x] Find task based on time and keyword
- [ ] Reminders (coming soon)

I am not sure you want to see coding anymore if it has tortured you so much that you need the app to remind you things, but this is the `main` method for the app:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

heeeyi committed:
Finished ip Level-1, added DukeEngine.java

heeeyi committed:
Add Task.java, modified Duke.java & DukeEngine.java

heeeyi committed:
added markDone() and markUnDone() in Task.java

heeeyi committed:
Added todo, deadline, event as subclasses

heeeyi committed:
Finished event, todo, deadline time detection

heeeyi committed:
Added input.txt, ACTUAL.txt, EXPECTED.txt; Add PAUSE to runtest.bat

heeeyi committed:
handle all current possible exceptions

heeeyi committed:
add deletions and corresponding exceptions

heeeyi committed:
put all the files in packages, set up unit testing

heeeyi committed:
Add hard disk Save function

Add generateTaskList(), fetchTask() in DukeEngine.java, add exception handlers in Duke.java

heeeyi committed:
Add Dates and Times

The engine now recognises deadline and event time as dates instead of strings.
Can use "search 01/03/2020" to search events due/happening at the date.

heeeyi committed:
Merge branch 'branch-level-7'

heeeyi committed:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/dukes/engine/Duke.java
#	src/main/java/dukes/engine/DukeEngine.java
#	src/main/java/dukes/engine/Task.java
#	src/main/java/dukes/engine/ToDo.java

Merge branch-level-8

heeeyi committed:
Fixed issues about Dates and Times

Successfully implemented dates and times function in Dukes,
making it compatible with the save function.

heeeyi committed:
Merge remote-tracking branch 'origin/add-gradle-support'

heeeyi committed:
Add support for gradle

Merge add-gradle-support branch into the main project, set up gradle to run tasks

heeeyi committed:
Implement more OOP structure

Refactor the code to extract classes UI, Storage, Parser, TaskList, Command and its subclasses.

heeeyi committed:
Organise into Packages

Organise the code into packages dukes.engine, dukes.command, dukes.task and dukes.util

heeeyi committed:
Add JUnit Tests

Set up Gradle to run JUnit tests, add tests for DeadLine.java, ToDo.java and Parser.java

heeeyi committed:
Package the app as a JAR file

Create the jar file containing the function till now for Duke

heeeyi committed:
Develop Find functionality

Allows user to search tasklist by keyword and find the respective tasks

heeeyi committed:
Add JavaDoc comments

Add JavaDoc comments for all classes and non-private methods,
except for getter and setter methods.

heeeyi committed:
Standardise the coding standard

Adjust the code to comply with the coding standard.

heeeyi committed:
Merge branch 'branch-Level-9'

heeeyi committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/dukes/util/TaskList.java
#	src/main/java/dukes/util/UI.java

Resolved

heeeyi committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/dukes/util/Parser.java
#	src/main/java/dukes/util/UI.java

Resolved

heeeyi committed:
Add Find feature, JavaDoc headers and coding standard

Adjust the merge and makes the duke software combine the 3 features of find, javadoc and coding standard

heeeyi committed:
Learn about basic UI for Duke

heeeyi committed:
Add a GUI to Duke through JavaFX

Using JavaFX and FXML to add a GUI to Duke, that allows user to directly input command and get response. Add Main, MainWindow and Launcher in dukes.engine

heeeyi committed:
Add JavaDoc to the new files

Add JavaDoc as method headers to the methods in MainWindow, Main, Launcher and new methods in Command family

heeeyi committed:
Merge branch 'branch-level-10'

heeeyi committed:
Add checkstyle to the project

Set up checkstyle for the coding using Gradle

heeeyi committed:
Merge branch 'branch-checkstyle'

PeanutButters93 commented on src/main/java/dukes/task/Task.java:
```diff
@@ -0,0 +1,95 @@
+        this.tag = "";
+    }
+
+    public void setDone() {
+        this.isDone = true;
+    }
+
+    public void setUnDone() {
+        this.isDone = false;
+    }
```
Perhaps consider combining these setter methods into 1?
public void setDone(boolean isDone)

PeanutButters93 commented on src/main/java/dukes/task/Task.java:
```diff
@@ -0,0 +1,95 @@
+        this.isDone = false;
+    }
+
+    public boolean hasDone() {
+        return this.isDone;
+    }
```
Function name fits convention but sounds grammatically weird. Might I suggest getDoneStatus()

PeanutButters93 commented on src/main/java/dukes/util/UI.java:
```diff
@@ -0,0 +1,235 @@
+public class UI {
+    // mainly use static methods
+    // need to handle the exceptions raised by Parser
+    public static final String WORD_DIVISHION_LINE = "________________________________________";
```
I think this might be a typo, might want to spell check division for clarity. But otherwise, I like the use of the constants. Very clean and neat looking.

PeanutButters93 commented on src/main/java/dukes/command/ListCommand.java:
```diff
@@ -0,0 +1,149 @@
+                    counter += 1;
+                    sb.append(counter).append(". ");
+                    sb.append(theTask.toString()).append("\n");
+                } else if (theTask.getTag().equals("E") &&
+                        (theTask.getStart().isBefore(date) || theTask.getStart().equals(date)) &&
+                        (theTask.getEnd().isAfter(date) || theTask.getEnd().equals(date))) {
```
This function has few comments and almost no line breaks. I find it a bit difficult to understand - especially this long boolean expression. Perhaps consider adding some line breaks and commenting this part?

PeanutButters93 reviewed (commented):
Looks v good and clean!
LGTM

FredericChow00 commented on src/main/java/dukes/engine/Duke.java:
```diff
@@ -0,0 +1,106 @@
+     * Runner of Duke class. Call classes to interpret command and execute.
+     * Catch all the runtime DukeExceptions and provide feedback.
+     */
+    public void run() {
```
I think since we are using the GUI and FXML for the handling of inputs, there is no longer a need for the use of Scanners since the front end of the GUI will call the handleUserInput() method in MainWindow.java. Perhaps this could be removed?

FredericChow00 commented on src/main/java/dukes/command/Command.java:
```diff
@@ -0,0 +1,91 @@
+     * @param storage handles the loading and saving of files.
+     * @throws DukeException if unexpected runtime issue occurs.
+     */
+    public abstract void execute(TaskList tasks, UI ui, Storage storage) throws DukeException;
```
Seems like `execute(...)` and `runCommand(...)` have very similar purposes. 

Imo you can just omit `execute(...)` since this is used in the `run()` method in Duke.java with Scanners etc but there isn't a need for that anymore with the GUI

FredericChow00 commented on src/main/java/dukes/engine/Duke.java:
```diff
@@ -0,0 +1,106 @@
+     * @param args keyboard arguments
+     */
+    public static void main(String[] args) {
+        new Duke("data/tasks.txt").run();
```
Perhaps you can replace this file path with the static class variable `DEFAULT_PATH`? Since it is already defined at the start

FredericChow00 reviewed (commented):
Good job! In general there is quite a substantial amount of OOP and abstraction such as having different classes for each type of Command. 

There are some parts of the codebase that could be ommitted though from what I am seeing, such as the use of Scanners to handle inputs. Other than that, everything else looks fine to me!

heeeyi committed:
Command classes : Refactor and clean some part of the code

Command class and its subclasses have some deprecated methods, such as
execute(), which is no longer used. This causes code redundancy.

This commit removes these codes.

The subclasses of Command also contain some long methods over 30 lines.
This reduces code readability and code quality.

For the duplicate code that is used in multiple subclasses, this commit
extracts them into public static methods in the superclass to be used in
 the subclasses, so as to reduce redundancy.

For the non-duplicate long methods within the subclasses, this commit
extracts them into private util methods in the subclasses. This
conforms to the OOP principle, and increases readability.

At the same time, this commit creates a new jar file, duke.jar, in
build/libs. The mainClassName in build.gradle is updated to be
“dukes.engine.Launcher”.

heeeyi committed:
Add Assertions to document assumptions

Add assertions for Parser.java to check the assumptions.

heeeyi committed:
Adjust code quality

Make further adjustments about code quality to refactor some methods

heeeyi committed:
Merge pull request #1 from heeeyi/branch-assertion

Add Assertions to document assumptions

heeeyi committed:
Merge branch 'master' of https://github.com/heeeyi/ip

heeeyi committed:
Merge branch 'master' into branch-code-quality

# Conflicts:
#	src/main/java/dukes/util/Parser.java
#	src/main/java/dukes/util/UI.java

Resolve conflicts

heeeyi committed:
Merge pull request #2 from heeeyi/branch-code-quality

Improve code quality

heeeyi committed:
Merge branch 'master' of https://github.com/heeeyi/ip

heeeyi committed:
Add the extension to update tasks

Now users can update an existing task in the following format:
* update 4 play games
* update 1 sleep /by 01/03/2022
* update 5 build /from 03/05/2023, etc.

heeeyi committed:
Merge branch 'branch-update'

heeeyi committed:
Add C-Update

heeeyi committed:
Set up CI, remove I/O redirection content

heeeyi committed:
Make GUI Adjustments

Improve the Duke GUI by:
* Add an icon to the application;
* Add background colour to the scroll pane and dialog box;
* Change the icon for Duke and user:
* Add username for user;
* Adjust size of the scroll bar.

heeeyi committed:
Add an image to showcase

heeeyi committed:
Rename the picture

heeeyi committed:
Update README.md as the UserGuide

heeeyi committed:
Some trial run

heeeyi committed:
Update README.md as the UserGuide v1.0

heeeyi committed:
Update README.md

heeeyi committed:
Update README.md as better user guide

heeeyi committed:
Update README.md

heeeyi committed:
Update README.md

heeeyi committed:
Update README.md

heeeyi committed:
Update README.md

heeeyi committed:
Update README.md

heeeyi committed:
Add "exit the app" method

heeeyi committed:
Update the README.md for exit the app

heeeyi committed:
Adjust styles
